sklearn/neighbors/_binary_tree.pxi

```diff
@@ -1049,18 +1049,17 @@ cdef class BinaryTree:
     def __init__(self, data,
                  leaf_size=40, metric='minkowski', sample_weight=None, **kwargs):
         # validate data
-        if data.size == 0:
+        self.data_arr = check_array(data, dtype=DTYPE, order='C')
+        if self.data_arr.size == 0:
             raise ValueError("X is an empty array")
 
+        n_samples = self.data_arr.shape[0]
+        n_features = self.data_arr.shape[1]
```

Comment:
```python
n_samples, n_features = self.data_arr.shape
```

Reply:
I have reverted to the original formulation. Cython was giving this error
```
            (n_samples, n_features) = self.data_arr.shape
                                                  ^
    ------------------------------------------------------------
    
    sklearn/neighbors/_binary_tree.pxi:1055:47: Cannot convert 'npy_intp *' to Python object
```
It has something to do with array conversion from numpy and cython, a bit subtle for me... like Cython is closer to C than to Python in this case?

Reply:
`np.ndarray` (data type of `self.data_arr`) is not an usual `numpy` array, but a `Cython` wrapper with [this signature](https://github.com/cython/cython/blob/master/Cython/Includes/numpy/__init__.pxd#L243-L295).

Note that the `shape` attribute is a pointer (`npy_intp *shape`). Therefore, tuple unpacking cannot be applied.

Reply:
Thanks @alfaro96! :)

Reply:
good point. It has been a while I did not review some Cython :)